feat: 커뮤니티 기능 추가 (KGU-13)

JinhyeokFang opened the pull request:
# 작업 내용

1. 커뮤니티에 필요한 API를 구현했습니다.

# 동작 화면

1.

# 리뷰 받고 싶은 부분

1. 테스트 코드에서 Mocking이 좀 어려운지라 그 부분이 테스트 목적에 맞게 작성되었는지 의견이 궁금합니다!

# 그 외 참고사항

1. 유저 인증 기능은 로그인 기능 추가 후 구현 예정입니다.

JinhyeokFang committed:
refactor: 코드 괄호 위치 수정 (KGU-6)

JinhyeokFang committed:
Merge pull request #8 from kgu-basic-capstone/feat/KGU-6-community-comment

feat: 커뮤니티 댓글 기능 (KGU-6)

JinhyeokFang committed:
Merge pull request #11 from kgu-basic-capstone/develop

Develop 브랜치 병합

JinhyeokFang committed:
feat: 커뮤니티 기능 추가 (KGU-13)

byb0823 commented on api/src/main/java/uk/jinhy/server/api/community/presentation/CommunityDto.java:
P2: 여러 DTO가 내부 클래스로 선언되어 있고, 그 안에 다시 내부 클래스가 있습니다. 하나의 클래스에 모든 DTO를 선언하는 것보다 개별적으로 분리하는 게 좋아보입니다. 내부 클래스의 내부 클래스에도 lombok이 필요 이상으로 사용되고 있는 것 같습니다. 필요한 것만 사용하도록 변경하는 것이 좋다고 생각합니다.

JinhyeokFang replied:
DTO 클래스를 어떻게 개별적으로 분리하는 것이 좋을까요?
저는 Application과 Presentation Layer에 각각 DTO가 존재한다면
계층별로 의존성도 분리되어 유지보수에 이점이 있을 거 같긴 한데
이 부분은 컨벤션으로 합의하고 넘어가야할 거 같아 우선은 Presentation 계층에만 몰아두긴 했습니다.
영빈님은 어떻게 생각하시나요?

byb0823 replied:
저도 각 layer별로 분리하는 것이 의존성 관리에 좋을 것 같다고 생각합니다.
다만 각 DTO 클래스가 하나의 자바 파일 내에 내부 클래스로 여러 개 선언되어 있는 것보다
개별적으로 선언하는 것이 좋다 생각되어 말씀드렸던 것이었습니다.
각 계층 디렉터리에 DTO 디렉터리를 개별적으로 만들어 관리하면 좋을 것 같다고 생각합니다.

JinhyeokFang replied:
그러면 우선은 계층별로 DTO 디렉토리를 만들어서 분리해두되
네이밍같은 세부 사항은 수요일 회의에서 픽스한 이후 수정하면 좋을 거 같습니다 :)
바로 반영해볼게요!

byb0823 commented on api/src/main/java/uk/jinhy/server/api/community/domain/CommunityPost.java:
```diff
@@ -0,0 +1,47 @@
+        return comment;
+    }
+
+    public void updateContent(String newContent) {
+        this.content = newContent;
+    }
+
+    public void update(String newTitle, String newContent, Category newCategory, User requestUser) {
+        if (!isAuthor(requestUser)) {
+            throw new IllegalArgumentException("게시글 수정 권한이 없습니다");
+        }
+        this.title = newTitle;
+        this.content = newContent;
+        this.category = newCategory;
+    }
```
P3: 업데이트 이후 성공/실패 여부를 반환해주면 테스트하기에 편할 것 같아요.

JinhyeokFang replied:
업데이트가 실패하는 경우 예외를 던지도록 구현했는데 더 부족한 부분이 있을까요?

(어떤 부분에서 부족함을 느꼈는지 말씀해주시면 더 도움될 거 같습니다!)

byb0823 replied:
이 부분은 제가 예외처리가 있는 것을 못봤습니다..
예외처리 정도면 충분한 것 같습니다!

byb0823 commented on api/src/main/java/uk/jinhy/server/api/domain/User.java:
```diff
@@ -1,26 +1,21 @@
+@AllArgsConstructor
+@Builder
 public class User {
 
     private Long id;
     private String username;
     private String email;
     private List<Pet> pets;
 
-    @Builder
-    private User(String username, String email) {
-        this.username = username;
-        this.email = email;
-        this.pets = new ArrayList<>();
-    }
-
```
P3: 생성자 레벨에 있는 builder 어노테이션은 생성자의 파라미터로 노출된 필드만 입력 받지만, 클래스 레벨에 있는 builder 어노테이션은 모든 파라미터를 입력 받습니다. builder 어노테이션을 클래스 레벨로 이동 시킨 이유가 궁금합니다.

JinhyeokFang replied:
이 부분도 밑의 NoArgsConstructor와 같은 이유로 변경된 부분이에요
Mocking을 위해 모든 프로퍼티를 set할 수 있도록 변경된 부분인데 이 부분도 어떻게 개선하는 게 좋을까요?

byb0823 replied:
이 부분도 아래 제가 남긴 답변과 같습니다!

byb0823 commented on service/src/main/java/uk/jinhy/server/service/community/application/CommunityServiceImpl.java:
P3: 도메인 엔티티와 JPA 엔티티를 분리한 이유가 궁금합니다.

JinhyeokFang replied:
JPA뿐만아니라 세부 사항에 의존적인 코드는 모두 service 모듈에서 구현하고
순수 비즈니스 로직는 api 모듈에서 정의함으로써 세부 사항에 비즈니스 로직이 의존되지 않도록 의도했습니다!
도메인 엔티티 또한 JPA라는 세부 사항으로부터 의존성을 가지지 않도록 분리하기 위해 POJO로 만들어졌는데요,
결국 Spring에 의존적이면 JPA에도 의존적이어도 괜찮다 판단할 수도 있다고 생각해요!

![image](https://github.com/user-attachments/assets/d3478f3a-19ce-4832-90b6-e33f8773c9cc)
저는 팀 개발 리소스가 충분해서 초기 비용이 부담스럽지 않기도 하고 장기적으로 봤을땐 유지보수에 이점이 있다고 생각해서
의존성을 분리하긴 했는데 영빈님의 생각도 궁금합니다!

byb0823 replied:
저도 POJO를 사용하여 특정 기술에 종속적이지 않게 개발하는 것은 유지보수에도 좋다고 생각합니다!

byb0823 commented on service/src/main/java/uk/jinhy/server/service/community/domain/CommunityPostEntity.java:
```diff
@@ -1,19 +1,19 @@
 
 @Entity
 @Getter
-@NoArgsConstructor(access = AccessLevel.PROTECTED)
```
P2: JPA엔티티는 기본 생성자가 필요한 것으로 아는데 제거한 이유가 궁금합니다.

JinhyeokFang replied:
NoArgsConstructor를 제거하지는 않았고 Protected에서 Public으로 제어자를 변경했습니다.
테스트를 작성할 때 JPA Entity를 Mocking하려면 생성자가 필요하다보니 제어자를 변경했는데
캡슐화가 깨진다는 문제가 있는지라 이 부분이 고민되네요...
캡슐화랑 테스트 코드를 둘 다 가져가고 싶은데 가져갈 수 있을지, 안된다면 어떻게 Tradeoff 해야할 까 고민이 되는데 어떻게 생각하시나요?

byb0823 replied:
Mocking의 단점은 객체의 캡슐화가 깨지고 리팩터링 내성이 떨어지는 데에 있습니다.
무조건 Mocking을 지양하자는 것은 아닙니다. 
하지만 꼭 필요한 곳에만 제한적으로 사용하는 것이 좋다 생각합니다.

단위 테스트에서 단위를 하나의 기능으로 정의한다면,
JPA Entity같은 경우 외부에서 명확하게 식별할 수 있는 기능이 아니기 때문에 세부 구현에 해당한다고 판단했습니다.

따라서 mocking을 하기보단 테스트용 정적 팩터리 메서드 등을 통해 실제 객체를 사용하면 캡슐화와 테스트 코드 간 균형을 잘 잡을 수 있다고 생각해요.

byb0823 commented on service/src/main/java/uk/jinhy/server/service/community/domain/CommunityPostRepository.java:
P2: 검색과 같은 동적 쿼리는 JPA보단 QueryDSL을 사용하는 것이 더 가독성에 좋습니다.

JinhyeokFang replied:
QueryDSL로 개선하면 확실히 가독성면에서 좋고 검색 조건을 추가하거나 제거하기도 좋겠네요
우선은 지금 요구사항과 쿼리 정도는 Production에 반영되어도 코드 품질이나 동작에 치명적이진 않은거 같아서
지금 상태로 반영하는 대신 QueryDSL로 개선하는 티켓을 따로 만들어서 작업하면 좋을 거 같은데 어떻게 생각하시나요?

(코드 리뷰 시에 왜 가독성이 좋을까? 에 대한 생각과 예시도 적어주신다면 리뷰이가 더 도움이 됩니다!)

byb0823 replied:
저도 지금 상태 그대로 반영해도 괜찮다고 생각합니다.

JPQL같은 경우 문자열로 작성되기 때문에 가독성이 상대적으로 떨어진다고 판단하였습니다.

추가적으로 문자열을 사용하기 때문에 컴파일 타임이 아닌 런타임에 에러를 발견할 수 있고
조건이 많아진다면 확장성이 떨어지기 때문에 QueryDSL을 사용하는 것이 좋다고 생각했습니다.
때문에 당장은 필요 없다고 생각하지만 조건문이 많아지면 그때 티켓을 만들어 개선하면 좋을 것 같아요!

byb0823 commented on service/src/test/java/uk/jinhy/server/service/community/application/CommunityServiceTest.java:
```diff
@@ -0,0 +1,637 @@
+@ExtendWith(MockitoExtension.class)
+class CommunityServiceTest {
+
+    @Mock
+    private CommunityPostRepository postRepository;
+
+    @Mock
+    private CommunityCommentRepository commentRepository;
```
P3: 리포지토리에 h2와 같은 테스트 db를 사용하는 방법이 아닌 mock을 사용한 이유가 궁금합니다.

JinhyeokFang replied:
우선은 작업 범위가 이미 많아서 Unit Test만 작성하다보니 실제 테스트 DB를 사용한 시나리오 테스트가 아쉽게도 빠지게 되었습니다.
코드의 안정성도 중요하긴 하지만 프로덕트 초기인 만큼 빠르게 비즈니스 로직을 추가되는 것도 중요하다 생각했고
유닛 테스트만 추가하는 것으로 trade off하는 대신 e2e 테스트를 추가하는 티켓을 따로 발행했습니다만
제 생각이 틀릴수도 있고 다른 분들의 의견이 궁금하기도 한데, 영빈님은 어떻게 생각하시나요?

byb0823 replied:
단위 테스트에서 단위의 해석에는 크게 두 가지 관점이 있는 것으로 알고 있습니다.
첫 번째는 비즈니스 기능 단위, 두 번째는 클래스 또는 메서드와 같은 작은 코드 조각입니다.
전자는 실제 객체를 사용하여 리팩터링 내성을 확보하고, 후자는 mocking을 통해 테스트 대상의 의존성을 mock 객체로 대체하여 테스트 대상의 경계를 명확히 할 수 있다는 장점이 있습니다.

현재 코드는 두 번째 관점에서 작성된 것으로 보여 질문을 드렸습니다.
프로덕트 초기에는 진혁님 방식대로 mocking을 통해 빠르게 테스트 하는 것도 괜찮다고 생각됩니다!

JinhyeokFang replied:
확실히 피처단위로 테스트되지는 않다보니 리팩터링 내성을 확보하긴 어려운 거 같네요
testcontainer를 활용한 실제 DB로 테스트하는 방식으로 코드를 보강해야할 거 같네요 :)

byb0823 commented on service/src/test/java/uk/jinhy/server/service/community/application/CommunityServiceTest.java:
```diff
@@ -0,0 +1,637 @@
+            when(postRepository.findByIdWithAuthorAndComments(postId)).thenReturn(Optional.empty());
+
+            // when & then
+            assertThrows(PostNotFoundException.class, () -> communityService.getPost(postId));
```
P2: junit 보다는 assertj를 사용하여 작성하는 게 더 가독성에 좋을 것 같습니다.

JinhyeokFang replied:
헉 이 부분 AssertJ가 아니라 JUnit으로 되어있네요
AssertJ로 통일하는게 유지보수하는데 확실히 좋을 거 같네요!
이 부분은 간단하니 빠르게 수정해볼게요

JinhyeokFang replied:
assertThrows를 assertJ의 assertThatExceptionOfType로 수정했고
수정하는 김에 BDD에 맞게 BDDMockito를 활용해 개선해봤습니다!

JinhyeokFang commented:
태스크 범위도 매우 넓고 새로운 코드도 많아 리뷰하느라 힘드셨을 텐데 정말 고생 많으셨습니다!
확실히 다른 팀원의 시각에서 코드를 바라보니 생각할 포인트도 더 많아지고 확실히 더 아키텍처에 대해 깊게 고민할 수 있어서 좋았습니다!
다만, Requested Change를 선택한 개괄적인 이유와 각 커멘트에서 영빈님이 그렇게 생각하신 이유, 리뷰이가 남긴 "리뷰 받고 싶은 부분"에 대한 영빈님의 생각 또한 얘기 나누어보았다면 리뷰이로서 조금 더 도움이 많이 되었을 거 같아요!
정말 고생 많으셨습니다 :)

JinhyeokFang committed:
refactor: 테스트 코드의 assertion 변경 (KGU-13)

JUnit의 assertThrows를 사용하는 대신 assertJ의 assertThatExceptionOfType을 사용하도록 수정하고
BDD에 맞게 assertion을 개선했습니다.

JinhyeokFang committed:
refactor: dto 모듈 계층 별 생성 및 분리 (KGU-13)

presentation 등 계층별로 dto를 분리하고
하나의 클래스로 되어있던 dto를 여러 클래스로 분리했습니다.

JinhyeokFang committed:
refactor: JPA Entity 생성자 변경 (KGU-13)

JPA Entity의 생성자를 캡슐화되도록 protected와 private으로 변경했습니다.
